### ISSUE TITLE
[Crash] Assertion failure assigning resource to static local resource

### ISSUE THREAD

**Issue by bob80905** [microsoft/DirectXShaderCompiler#8580]
```hlsl
RWByteAddressBuffer gBuf0 : register(u0);

uint Pass_StaticLocal(uint idx) {
    static RWByteAddressBuffer buf = gBuf0;
    buf.Store(idx * 4, 1);

    return 1;
}

[numthreads(1,1,1)]
void main(uint3 tid : SV_DispatchThreadID) {
    Pass_StaticLocal(tid.x);
}
```
https://godbolt.org/z/dv4718TeW
There is an assertion failure of 
```
Internal Compiler error: cast<X>() argument of incompatible type!

Compiler returned: 29
```
When compiling and codegen target is DXIL, but it does not trigger when the codegen target is SPIRV.
